app/javascript/components/widget-display.jsx

```diff
@@ -0,0 +1,225 @@
+import React, { useRef, useEffect, useReducer } from 'react';
+import PropTypes from 'prop-types';
+import styled from 'styled-components';
+import { mathMod } from 'ramda';
+import { DisconnectedMessage, LoadingMessage } from '@messages/message';
+import FullPanel from '@components/full-panel';
+import SidePanel from '@components/side-panel';
+import Twitter from '@widgets/twitter';
+import Numbers from '@widgets/numbers';
+import Weather from '@widgets/weather';
+import Traffic from '@widgets/traffic';
+
+const Wrapper = styled.div`
+  outline: none;
+`;
+
+const widgetElements = {
+  Weather,
+  Twitter,
+  Numbers,
+  Traffic,
+};
+
+const prefetchDelay = 1500;
+let lastPrefetchTime = 0;
+
+const stateSwitch = (state, { type, changeValue }) => {
+  const { timerValue, pausedTime } = state;
+  switch (type) {
+    case 'setPausedTime':
+      return {
+        timerValue,
+        pausedTime: changeValue,
+      };
+    case 'extendTimer':
+      return {
+        timerValue: timerValue + changeValue,
+        pausedTime: null,
+      };
+    case 'resetTimerValue':
+      return {
+        timerValue: changeValue,
+        pausedTime,
+      };
+    default:
+      return state;
+  }
+};
+
+export const WidgetDisplay = ({
+  requestPrefetch,
+  requestWidget,
+  cityName,
+  widgets,
+}) => {
+  const { byIdOrFirst: current, enabled: enabledWidgets } = widgets;
+
+  const WidgetElement = widgetElements[current.name].Panel;
+
+  const index = enabledWidgets.findIndex(w => w.id === current.id);
+
+  const prevWidgetId =
+    enabledWidgets[mathMod(index - 1, enabledWidgets.length)].id;
+  const nextWidgetId =
+    enabledWidgets[mathMod(index + 1, enabledWidgets.length)].id;
+
+  const [state, dispatch] = useReducer(stateSwitch, {
+    timerValue: Date.now() + current.durationSeconds * 1000,
+    pausedTime: null,
+  });
+
+  const timerId = useRef(null);
+
+  const nextStateFromKey = ({ keyCode }) => {
+    const { pausedTime } = state;
+    const upKey = 38;
+    const downKey = 40;
+    const SPACEBAR = 32;
+    if (keyCode === SPACEBAR) {
+      if (pausedTime) {
+        dispatch({
+          type: 'extendTimer',
+          changeValue: Date.now() - pausedTime,
+        });
+      } else {
+        dispatch({ type: 'setPausedTime', changeValue: Date.now() });
+      }
+    }
```

Comment:
if we're dispatching differently based on whats in state, its a sign that that this calculation should be inside the reducer. I think the action you want here is `togglePause`

```js
if (keyCode === SPACEBAR) {
  dispatch({type: 'togglePause'})
}
```